samza-api/src/main/java/org/apache/samza/operators/descriptors/base/stream/InputDescriptor.java

```diff
@@ -166,22 +183,31 @@ public SubClass withDeleteCommittedMessages(boolean deleteCommittedMessages) {
   }
 
   @Override
-  public Map<String, String> toConfig() {
-    HashMap<String, String> configs = new HashMap<>(super.toConfig());
+  public Map<String, String> toConfig(Map<String, String> currentConfig) {
+    HashMap<String, String> generatedConfig = new HashMap<>(super.toConfig(currentConfig));
     String streamId = getStreamId();
     this.offsetDefaultOptional.ifPresent(od ->
-        configs.put(String.format(OFFSET_DEFAULT_CONFIG_KEY, streamId), od.name().toLowerCase()));
+        generatedConfig.put(String.format(OFFSET_DEFAULT_CONFIG_KEY, streamId), od.name().toLowerCase()));
     this.resetOffsetOptional.ifPresent(resetOffset ->
-        configs.put(String.format(RESET_OFFSET_CONFIG_KEY, streamId), Boolean.toString(resetOffset)));
+        generatedConfig.put(String.format(RESET_OFFSET_CONFIG_KEY, streamId), Boolean.toString(resetOffset)));
     this.priorityOptional.ifPresent(priority ->
-        configs.put(String.format(PRIORITY_CONFIG_KEY, streamId), Integer.toString(priority)));
+        generatedConfig.put(String.format(PRIORITY_CONFIG_KEY, streamId), Integer.toString(priority)));
     this.isBootstrapOptional.ifPresent(bootstrap ->
-        configs.put(String.format(BOOTSTRAP_CONFIG_KEY, streamId), Boolean.toString(bootstrap)));
+        generatedConfig.put(String.format(BOOTSTRAP_CONFIG_KEY, streamId), Boolean.toString(bootstrap)));
     this.isBoundedOptional.ifPresent(bounded ->
-        configs.put(String.format(BOUNDED_CONFIG_KEY, streamId), Boolean.toString(bounded)));
+        generatedConfig.put(String.format(BOUNDED_CONFIG_KEY, streamId), Boolean.toString(bounded)));
     this.deleteCommittedMessagesOptional.ifPresent(deleteCommittedMessages ->
-        configs.put(String.format(DELETE_COMMITTED_MESSAGES_CONFIG_KEY, streamId),
+        generatedConfig.put(String.format(DELETE_COMMITTED_MESSAGES_CONFIG_KEY, streamId),
             Boolean.toString(deleteCommittedMessages)));
-    return Collections.unmodifiableMap(configs);
+    this.broadcastPartitionsOptional.ifPresent(broadcastPartitions -> {
```

Comment:
It seems like it might get confusing for toConfig to return a "global" value, and it would be simpler if toConfig was independent of existing configs.
Would it be possible to manage this through the graph and then generate the full "task.broadcast.inputs" like how "task.inputs" gets generated (I think it happens in JobNode)? I'm not really sure what it would look like, when initially seeing this, it seemed like "task.broadcast.inputs" should be treated more similarly to "task.inputs".

Reply:
Agree that this pattern is confusing and error prone. However there's currently no single class and pattern for all config generation. It's done by a combination ExectionPlanner, ApplicationDescriptorImpl and JobNode, StreamDescriptors and TableSpec. We do want to consolidate and simplify this logic. 

We can either check this in now and fix the internal APIs as part of one of SAMZA-1811, 1814 or 1842, or I can try to pass this around through the ApplicationDescriptorImpl. I'd prefer the former, since I'm not sure the latter is going to be the right way after the eventual refactor. WDYT?

Reply:
Would it be feasible to pass it through the StreamDescriptor through StreamGraphSpec until it reaches JobNode, and then create the config there (kind of like how "task.inputs" is generated)?
Do you think having the currentConfig argument in the toConfig signature is a necessary long-term pattern? To me, it seems like it might not be necessary. If the toConfig signature makes it into a release with the currentConfig argument, and then it will be backwards incompatible to remove it (since it seems to be part of the public API for stream descriptor implementors).

Reply:
These need to be generated in ExecutionPlanner, before the JobNode. I'll pass it around through StreamGraphSpec/ApplicationDescriptorImpl for now.

Reply:
I agree with Cameron on this point to move the generation of task.broadcast.inputs to JobNode, similar to task.inputs. We do have a fragmentation of configuration generation code across multiple stages of the planner: JobPlanner creating the system and stream configuration (and potential override of system and streams) to create the StreamManager, and then the ExecutionPlanner. The ExecutionPlanner creates JobGraph assuming the system stream configuration are already in the configure and use it to generate StreamSpec to create StreamEdge and JobNode. Then, finally JobNode generates the configuration for tables, stores, sideInputs, and serdes. It would be much cleaner if we can just use system, stream, store, and table descriptors in the planners and only use the descriptors to generate configuration and apply overrides in JobNode. This would require the following refactoring:
1. make sure that we don't rely on configuration to create StreamManager, ExecutionPlanner, JobGraph, StreamEdge, and JobNode
2. configuration is only generated from descriptors in JobNodeConfigureGenerator (in my PR, I separated the logic to generate configuration for JobNode to JobNodeConfigureGenerator).

In this PR, I would still prefer to move the generation of task.broadcast.inputs to JobNode. I don't see a need to use the configuration task.broadcast.inputs in ExecutionPlanner and JobGraph.